### PR TITLE
Make it clear that code outside plugins/ can have other free software licenses as well

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Right now, the rules state that *"[a]ll other code must be under the GPL-3.0-or-later"*. From the explanation it is clear that this sentence is about all other code in plugins/, but it is not formulated that way.

The current formulation implies for example that all unit tests, helper scripts, or code files included in roles, playbooks, or integration tests must be licensed under GPLv3+. Especially for unit tests this can be problematic; for example community.docker now vendors some code from Docker SDK for Python (Apache 2.0 licensed) including unit tests for that code. The unit tests are also licensed under Apache 2.0, but this would violate a strict interpretation of the inclusion requirements.

This PR proposes an improved formulation that makes it clear that other code **outside `plugins/`** can also use other licenses.

(Ref: https://github.com/ansible-collections/community.docker/pull/433#discussion_r933993684)

##### ISSUE TYPE
- Bugfix Pull Request
- Docs Pull Request

##### COMPONENT NAME
inclusion requirements

### CODE REVIEW

File: collection_requirements.rst
Comment (gotmax23) [inline collection_requirements.rst:269]: This does not seem to be formatted correctly. The indentation is really wonky in the rendered version.

**Comment by gotmax23** [comment]
I agree with the premise of this change. Thank you for proposing it, @felixfontein! I think `All other code` should be clarified to something like `All other plugins` and then the order of rows 3 and 4 should be switched.

File: collection_requirements.rst
Comment (felixfontein) [reply collection_requirements.rst]: Fixed in 964631699a266c6ccd3e4348b2d164a2a3a91e45.

**Comment by felixfontein** [comment]
> I agree with the premise of this change. Thank you for proposing it, @felixfontein! I think `All other code` should be clarified to something like `All other plugins` and then the order of rows 3 and 4 should be switched.

I implemented that in 3c36d09a90e92b9760b2c841e6d4e344148fe7a3.

File: collection_requirements.rst
Comment (gotmax23) [inline collection_requirements.rst:270]: Sorry for not batching up my feedback so you could apply it all at once.

```suggestion
:All other code: Code outside ``plugins/``, in particular unit tests for ``modules`` and ``module_utils``, must be licensed with
```

Comment (felixfontein) [reply collection_requirements.rst:270]: No problem about batching :)

I'm not sure why unit tests for modules and module_utils should be singled out here. Other unit tests also have to be licensed with a free software license (though very likely with GPLv3+, at least the way I interpret it, since IANAL).

Comment (gotmax23) [reply collection_requirements.rst:270]: Well, wouldn't the tests for the other plugins import the code from the other plugins that have to be GPL-licensed?

Comment (felixfontein) [reply collection_requirements.rst:270]: Yes, but (there's always a but ;) ) you can also have Python files there that do not import non-module/module_utils plugins. For example you could have a Apache 2.0 licensed file that provides some utility that is only used by a test for an inventory plugin, and thus located right next to it. Or you can have generic unit test helpers in `tests/unit/` outside of `tests/unit/plugins/modules/` and `tests/unit/plugins/module_utils/` (like the mock import helper) that are licensed under another license than GPLv3+.

Comment (gotmax23) [reply collection_requirements.rst:270]: Maybe it makes sense to phrase it like so?

```
:All other code: Code outside ``plugins/``, in particular unit tests [for ``modules`` and ``module_utils``], may be licensed under
                 another free software license that is compatible with the
                 `GPL-3.0-or-later <https://www.gnu.org/licenses/gpl-3.0-standalone.html>`_,
                 provided that such code that does not import any code that is licensed under
                 the ``GPL-3.0-or-later``.
```

(The bracketed part would either be removed or included without the brackets)

I don't feel strongly either way, but I think my suggestion might be clearer.

**Comment by gotmax23** [comment]
What is the process for getting this approved? Can the SC just vote on it or does it have to go through legal?

**Comment by felixfontein** [comment]
> What is the process for getting this approved? Can the SC just vote on it or does it have to go through legal?

I think a SC vote suffices, since this doesn't really change much. (Assuming nobody from SC disagrees and asks for a review by RH legal... I really don't think that will happen though.)

File: collection_requirements.rst
Comment (felixfontein) [reply collection_requirements.rst:270]: Hmm, I think this is only easier to understand for folks who know that code importing GPLv3+ licensed code needs to be GPLv3+ licensed as well. How about adding something like "Code satisfying the latter must be licensed under the `GPL-3.0-or-later`."?

Comment (gotmax23) [reply collection_requirements.rst:270]: I've edited the suggestion.

Comment (felixfontein) [reply collection_requirements.rst]: Applied as 0a1b0f352485e1ddbbac944130499b2b2a205440 (without the square brackets).

File: collection_requirements.rst
Comment (gotmax23) [inline collection_requirements.rst:268]:
```suggestion
                                 the ``GPL-3.0-or-later`` and often must import code from the controller.  For these
```

**Comment by gotmax23** [comment]
Besides that one other change, this looks good to me.

**Review by briantist — APPROVED** [review]

File: collection_requirements.rst
Comment (samccann) [inline collection_requirements.rst:273]:
```suggestion
                 provided that such code  does not import any other code that is licensed under
```

**Review by Andersson007 — APPROVED** [review]

File: collection_requirements.rst
Comment (mattclay) [reply collection_requirements.rst:273]: Since unit tests import modules, plugins or ansible-core code, it's probably worth clarifying that here.

**Comment by mattclay** [comment]
Since unit tests import modules, plugins or ansible-core code, it's probably worth clarifying that.

[repeating since my original statement was attached to a now hidden comment]

**Review by samccann — APPROVED** [review]
LGTM

File: collection_requirements.rst
Comment (mattclay) [inline collection_requirements.rst:270]: This section is long enough that I'm concerned someone will see the "in particular unit tests" part and not notice the caveat at the end regarding unit tests. Perhaps it would be adequate to remove the "in particular unit tests" part of the sentence?

Comment (mattclay) [reply collection_requirements.rst:270]: Or perhaps unit tests should have their own section ahead of "all other code" so we can more clearly address the topic there without making this section longer.

Comment (felixfontein) [reply collection_requirements.rst:270]: How about rewriting this section as
```
:All other code: Code outside ``plugins/`` may be licensed under another free software license that is compatible
                 with the `GPL-3.0-or-later <https://www.gnu.org/licenses/gpl-3.0-standalone.html>`_,
                 provided that such code does not import any other code that is licensed under
                 the ``GPL-3.0-or-later``. If the file does import other ``GPL-3.0-or-later`` code,
                 then it must similarly be licensed under ``GPL-3.0-or-later``. Note that this applies in
                 particular to unit tests; these often import code from ansible-core, plugins, module utils,
                 or modules, and such code is often licensed under ``GPL-3.0-or-later``.
```

Comment (mattclay) [reply collection_requirements.rst:270]: I think that would be adequate.

Comment (felixfontein) [reply collection_requirements.rst]: I've added it in 626b67e21400d9670308e99b4535277662025eb7. Thanks!

**Comment by felixfontein** [comment]
Any more comments? Or should we start voting on the current state?

**Comment by felixfontein** [comment]
There is now an active vote on this topic: https://github.com/ansible-community/community-topics/discussions/142

File: collection_requirements.rst
Comment (gotmax23) [inline collection_requirements.rst:269]: `s/GPLv3+/GPL-3.0-or-later/`

**Comment by felixfontein** [comment]
(I rebased after #216 was merged, no content changes.)

File: collection_requirements.rst
Comment (felixfontein) [reply collection_requirements.rst]: Implemented in 48da08635ce9bdf7a571572610f17f3dbaae446b.

**Review by Andersson007 — COMMENTED** [review]
@felixfontein thanks for the clarifications!
I have 2 questions:
1. Do i understand correctly that if a collection is fully licensed with a GPL compatible license and contains, say, `plugins/actions/..` or anything else besides `modules` and `module_utils` in the `plugins` directory, the maintainers must re-license it with GPLv3+, otherwise we can't include/have such a collection included?
2. Should we use uppercase modal verbs as we do in other paragraphs?

Thanks

**Comment by felixfontein** [comment]
>     1. Do i understand correctly that if a collection is fully licensed with a GPL compatible license and contains, say, `plugins/actions/..` or anything else besides `modules` and `module_utils` in the `plugins` directory, the maintainers must re-license it with GPLv3+, otherwise we can't include/have such a collection included?

Yes. But that's not related to including the collection into Ansible, it's a result of the GPL already. Action plugins (and other plugin types, except a few of them) import code from ansible-core, and thus these plugins **must** be licensed under GPLv3 as well.

**Comment by felixfontein** [comment]
> Should we use uppercase modal verbs as we do in other paragraphs?

Hmmm, I'm not sure whether this is a good idea to change since the vote already started.

**Comment by Andersson007** [comment]
> > Should we use uppercase modal verbs as we do in other paragraphs?
> 
> Hmmm, I'm not sure whether this is a good idea to change since the vote already started.

It doesn't change the meaning, just makes it a bit more recognizable (anyway it's not an obstacle to merge it as-is and we can do it later in another PR). Let's continue without changing it here

**Comment by felixfontein** [comment]
This PR has been approved, see https://github.com/ansible-community/community-topics/issues/123#issuecomment-1275020592.

**Comment by felixfontein** [comment]
Thanks everyone! :)

**Pull request merged**
